File: problem_30/digit_fifth_powers_test.rb
require 'minitest/autorun'
require 'minitest/pride'

class DigitFifthPowersTest < Minitest::Test
  def digit_power_sum(power)
    sum = 0
    range = 10..(10 ** (power + 1))

    range.each do |num|
      digits = num.to_s.chars.map { |dig| dig.to_i }
      local_sum = digits.sum { |dig| dig ** power}
      sum += local_sum if local_sum == num
    end

    sum
  end

  #----------------- Tests -----------------#

  def test_digit_power_sum
    assert_equal 19316,  digit_power_sum(4)
    assert_equal 443839, digit_power_sum(5)
  end
end

File: problem_30/digit-fifth-powers.test.js
function digitPowerSum(power) {
  let sum = 0;

  for (n = 10; n < Math.pow(10, power + 1); n++) {
    let digits = [...n.toString()].map((dig) => parseInt(dig));
    let localSum = 0;

    digits.forEach((dig) => (localSum += Math.pow(dig, power)));

    if (localSum === n) {
      sum += n;
    }
  }

  return sum;
}

test("digitPowerSum", () => {
  expect(digitPowerSum(4)).toBe(19316);
  expect(digitPowerSum(5)).toBe(443839);
});
